### PR TITLE
Skinny Deferred

### PR DESCRIPTION
**Pull request by zwaldowski**
`Deferred` is meant to handle large parts of an app's asynchronous control flow, so its performance is critical. We've strived in the past to use the best constructs we have at our disposal to make this possible, like being early adopters of GCD in Swift. #213 revealed we can still do better! 😃

### What's in this pull request?

By combining our waiters queue and the storage, `Deferred` (the type) can reach same-ballpark performance as GCD while maintaining the same invariants.

Per the performance tests, this PR features:

- A 70% reduction in peak memory usage
- A 35% speed improvement in simple operations
- An 80% speed improvement in chained operations
- Smaller call stacks when debugging `upon` handlers.

In summary: same great taste, less filling.

This representation allows Swift-side access to handlers, possibly allowing for #203.

#### TODO

- [x] What about code size? This PR gains about 30KB for Release under 4.1, which is livable, but not ideal for my OCD.

### Testing

Largely no impact, but `Deferred` gains a specialization `where Value: AnyObject` that needs testing.

#### TODO

- [x] Needs duplicated tests to handle `AnyObject` specialization
- [x] Double-check other coverage implications

### API Changes

No public API changes here, it's all under the hood.

### CODE REVIEW

File: Sources/Deferred/Deferred.swift [hunk 1]
```diff
@@ -8,133 +8,77 @@
...
+        } ?? handler(value)
     }
+}
 
-    static func convertFromReference(_ value: AnyObject) -> Value {
-        return unsafeDowncast(value, to: Box.self).wrapped
+extension Deferred: PromiseProtocol {
+    @discardableResult
+    public func fill(with value: Value) -> Bool {
+        return variant.awake(with: value)
```
Comment (zwaldowski) [inline Sources/Deferred/Deferred.swift]: Rename `store(_:)`.
